src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -95,56 +95,77 @@ EnsureReferenceTablesExistOnAllNodes(void)
 	 *
-	 * We also don't need special handling for concurrent create_refernece_table.
-	 * Since that will trigger a call to this function from another backend,
-	 * which will block until our call is finished.
+	 * Since the changes to the reference table placements are made via loopback
+	 * connections we release the final lock held at the end of this function.
 	 */
-	int colocationId = CreateReferenceTableColocationId();
-	LockColocationId(colocationId, ExclusiveLock);
-
-	List *referenceTableIdList = CitusTableTypeIdList(REFERENCE_TABLE);
-	if (referenceTableIdList == NIL)
+	LOCKMODE lockmodes[] = { AccessShareLock, ExclusiveLock };
+	for (int l = 0; l < lengthof(lockmodes); l++)
 	{
-		/* no reference tables exist */
-		UnlockColocationId(colocationId, ExclusiveLock);
-		return;
-	}
+		LockColocationId(colocationId, lockmodes[l]);
 
-	Oid referenceTableId = linitial_oid(referenceTableIdList);
-	const char *referenceTableName = get_rel_name(referenceTableId);
-	List *shardIntervalList = LoadShardIntervalList(referenceTableId);
-	if (list_length(shardIntervalList) == 0)
-	{
-		/* check for corrupt metadata */
-		ereport(ERROR, (errmsg("reference table \"%s\" does not have a shard",
-							   referenceTableName)));
-	}
+		referenceTableIdList = CitusTableTypeIdList(REFERENCE_TABLE);
+		if (referenceTableIdList == NIL)
+		{
+			/* no reference tables exist */
+			for (int ll = l; ll >= 0; ll--)
+			{
+				UnlockColocationId(colocationId, lockmodes[ll]);
+			}
+			return;
+		}
 
-	ShardInterval *shardInterval = (ShardInterval *) linitial(shardIntervalList);
-	uint64 shardId = shardInterval->shardId;
+		Oid referenceTableId = linitial_oid(referenceTableIdList);
+		referenceTableName = get_rel_name(referenceTableId);
+		List *shardIntervalList = LoadShardIntervalList(referenceTableId);
+		if (list_length(shardIntervalList) == 0)
+		{
+			/* check for corrupt metadata */
+			ereport(ERROR, (errmsg("reference table \"%s\" does not have a shard",
+								   referenceTableName)));
+		}
 
-	/*
-	 * We only take an access share lock, otherwise we'll hold up citus_add_node.
-	 * In case of create_reference_table() where we don't want concurrent writes
-	 * to pg_dist_node, we have already acquired ShareLock on pg_dist_node.
-	 */
-	List *newWorkersList = WorkersWithoutReferenceTablePlacement(shardId,
-																 AccessShareLock);
-	if (list_length(newWorkersList) == 0)
-	{
-		/* nothing to do, no need for lock */
-		UnlockColocationId(colocationId, ExclusiveLock);
-		return;
+		ShardInterval *shardInterval = (ShardInterval *) linitial(shardIntervalList);
+		shardId = shardInterval->shardId;
+
+		/*
+		 * We only take an access share lock, otherwise we'll hold up citus_add_node.
+		 * In case of create_reference_table() where we don't want concurrent writes
+		 * to pg_dist_node, we have already acquired ShareLock on pg_dist_node.
+		 */
+		newWorkersList = WorkersWithoutReferenceTablePlacement(shardId, AccessShareLock);
```

Comment:
Hmm, I get some occasional deadlocks with concurrent `DROP TABLE ref` and citus_add_node(). Is that expected?


```SQL
ERROR:  deadlock detected
DETAIL:  Process 87658 waits for RowShareLock on relation 20052 of database 13236; blocked by process 87605.
Process 87605 waits for AccessShareLock on relation 213392 of database 13236; blocked by process 87658.
HINT:  See server log for query details.
CONTEXT:  SQL statement "SELECT master_remove_distributed_table_metadata_from_workers(v_obj.objid, v_obj.schema_name, v_obj.object_name)"
PL/pgSQL function citus_drop_trigger() line 13 at PERFORM
```

Where:
```SQL
select 20052::regclass, 213392::regclass;
┌──────────────┬─────────────────┐
│   regclass   │    regclass     │
├──────────────┼─────────────────┤
│ pg_dist_node │ gexec_test_1613 │
└──────────────┴─────────────────┘
(1 row)


```

Reply:
deadlocks sound not good, how reliable do you get these? What surprises me is that they deadlock on relation locks. These I didn't change. However, it could be that due to allowing concurrent reference table drops this deadlock shows up now. Do you still have the database around?

What are the table names for relation `213392` and `20052`?

Does this reproduce on `main`?
I think this might be related to a separate thing I noticed while going through the code.

We take a lower level lock on the table being dropped in the `preprocess` hook, then obtain the colocation id lock before postgres tries to acquire a high lock on the table to be dropped. I was thinking to directly take the high lock in the preprocess hook for dropping a table. To reduce the deadlock potential.

If this doesn't reproduce on `main` we need to understand the lock graph better with regards to the lock changes I made.

It will probably not deadlock on reference table drop's as those are serialized on `main` on the colocation id lock. But distributed tables could have the same behaviour you are observing.

Reply:
> What are the table names for relation 213392 and 20052?

They are on the comment:  pg_dist_node and gexec_test_1613 

> Does this reproduce on main?

oh, yes it seems so.  I can reliably repro on both main and this branch, so most likely unrelated to this PR.

Is it probably fine to open issue for now?